File: news/moneycontrol.js
// Senior Developer: ChatGPT
// Intern: Satyajit
require('dotenv').config();
const puppeteer = require('puppeteer');
const mongoose = require('mongoose');
const companies = require('../companyList'); 

const MONGODB_URI = process.env.MONGODB_URI;

mongoose.connect(MONGODB_URI, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
}).then(() => {
  console.log('Connected to MongoDB');
}).catch(error => {
  console.error('Error connecting to MongoDB:', error);
});

const News = mongoose.model('news', new mongoose.Schema({
  headline: String,
  link: String,
  datetime: String,
  imageUrl: String,
  tag: String, // Add tag field to the schema
}));

async function scrapeMoneyControlNews() {
  const browser = await puppeteer.launch({ headless: false });
  const page = await browser.newPage();
  const url = 'https://www.moneycontrol.com/news/tags/companies.html';
  
  try {
    await page.goto(url);

    // Wait for the news list to load (you may need to adjust the selector or wait time)
    await page.waitForSelector('.clearfix', { timeout: 10000 });

    const newsData = await page.evaluate(() => {
      const newsList = Array.from(document.querySelectorAll('.clearfix'));

      return newsList.map((newsItem) => {
        const headlineElement = newsItem.querySelector('h2 a');
        const linkElement = newsItem.querySelector('h2 a');
        const datetimeElement = newsItem.querySelector('span');
        const imageElement = newsItem.querySelector('img');

        const headline = headlineElement ? headlineElement.innerText : null;
        const link = linkElement ? linkElement.href : null;
        const datetime = datetimeElement ? datetimeElement.innerText : null;
        const imageUrl = imageElement ? imageElement.getAttribute('data-src') : null;

        return { headline, link, datetime, imageUrl };
      });
    });

    const validNewsData = newsData.filter(item => item.headline && item.link && item.datetime && item.imageUrl);

    return validNewsData;
  } catch (error) {
    console.error('Error:', error);
  } finally {
    await browser.close();
  }
}

async function saveMoneyControlNewsToDB(newsArr) {
  try {
    for (const element of newsArr) {
      try {
        const existingNews = await News.findOne({ headline: element.headline });

        if (!existingNews) {
          const matchingCompany = companies.find(company => element.headline.includes(company));
          element.tag = matchingCompany ? matchingCompany : '';

          const newsItem = new News(element);

          await newsItem.save();
          console.log('News article added successfully.');
        } else {
          console.log('News article already exists. Skipping.');
        }
      } catch (err) {
        console.error('An error occurred:', err);
      }
    }
  } catch (error) {
    console.error('Error connecting to MongoDB:', error);
  }
}

(async () => {
  try {
    const newsData = await scrapeMoneyControlNews();

    console.log('Extracted Data:', newsData);
    console.log('Total News:', newsData.length);

    saveMoneyControlNewsToDB(newsData);
  } catch (error) {
    console.error('Error:', error);
  }
})();

File: news/moneycontrol.py

from time import sleep

import requests
from bs4 import BeautifulSoup
from geminiapi import generate_summary


def save_to_mongodb(news_data):
    url = "http://localhost:5000/api/v1/news"
    headers = {"Content-Type": "application/json"}

    for news_item in news_data:
        try:        
            response = requests.post(url, json=news_item, headers=headers)
            if response.status_code == 201:
                print(f'News article "{news_item["title"]}" added successfully.')
            elif response.status_code == 409:  # Assuming 409 Conflict is returned for duplicates
                print(f'News article "{news_item["title"]}" already exists. Skipping.')
            else:
                print(f'Failed to add news article "{news_item["title"]}". Status code: {response}')
        except Exception as e:
            print(f'Error occurred while adding news article "{news_item["title"]}": {str(e)}')

def get_article_content(article_url, headers):
    response = requests.get(article_url, headers=headers)
    
    if response.status_code != 200:
        print(f"Error: Unable to access article page (Status code: {response.status_code})")
        return None

    soup = BeautifulSoup(response.content, 'lxml')
    # Adjust the selector based on the actual structure of the article pages
    article_content_element =soup.find('div', class_='content_wrapper')   # Replace with the actual class/id of the article content
    
    article_content = article_content_element.get_text(strip=True) if article_content_element else None
    return article_content

def scrape_money_control_news():
    url_iteration = 0
    # there is a bug where second url get skipped
    urls = {'https://www.moneycontrol.com/news/tags/companies.html','https://www.moneycontrol.com/news/business/economy/','https://www.moneycontrol.com/news/business/mutual-funds/', 'https://www.moneycontrol.com/news/business/personal-finance/', 'https://www.moneycontrol.com/news/business/ipo/', 'https://www.moneycontrol.com/news/business/startups/','https://www.moneycontrol.com/news/business/stocks/'}
    # urls = {'https://www.moneycontrol.com/news/business/personal-finance/',}
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    news_data = []
    for url in urls:
        url_iteration += 1
        print(url_iteration)
        response = requests.get(url, headers=headers)
       
        if response.status_code != 200:
            print(f"Error: Unable to access page (Status code: {response.status_code})")
            return []

        soup = BeautifulSoup(response.content, 'lxml')
        news_items = soup.select('.clearfix')

        
        item_iteration = 0
        for item in news_items:
            
            inner_text_element = item.select_one('h2 a')
            news_time_element = item.select_one('span')
            image_element = item.select_one('img')
            lazy_element = item.select_one('.lazy-container')

            inner_text = inner_text_element.get_text(strip=True) if inner_text_element else None
            href = inner_text_element['href'] if inner_text_element and 'href' in inner_text_element.attrs else None
            news_time = news_time_element.get_text(strip=True) if news_time_element else None
            image_url = image_element['data-src'] if image_element and 'data-src' in image_element.attrs else None
            if news_time == "Jump to" and lazy_element:
                news_time = lazy_element.select_one('span')
                news_time = news_time.get_text(strip=True) if news_time_element else None
            # Extract title from inner_text_element if available    
            title = inner_text_element['title'] if inner_text_element and 'title' in inner_text_element.attrs else None

            if inner_text and href and news_time and image_url:
                article_url = href
                article_content = get_article_content(article_url, headers)
                sleep(1)  # To avoid making too many requests in a short period
                result = generate_summary(title=title,content=article_content)
                if result == None:
                    print(article_url)
            
                item_iteration += 1
                print(url_iteration,":",item_iteration)

                news_data.append({
                    'title': title,
                    'summary':result["summary"] if result != None else "None",
                    'articleUrl': article_url,      
                    'category':result["categories"] if result != None else "None",
                    'company':result["company_name"][0:2] if result != None and result.get("company_name")  else "None",           
                    'tags':result["tags"] if result != None else "None",
                    'publishTime': news_time.replace(' IST', ''),
                    'imageUrl': image_url,
                })

    return news_data



if __name__ == '__main__':
        news_data = scrape_money_control_news()

        if news_data:
            filtered_data = [item for item in news_data if item.get("summary") is not None and item.get("summary") != "None" ]
            print('Extracted Data:', filtered_data)
            print('Total News:', len(filtered_data))
        
            save_to_mongodb(news_data)
            print('News data saved to MongoDB.')
        else:
            print('No news data extracted.')

